joomla/joomla-cms
# [4.0] [com_users] Fix default value for datetime columns and make some nullable

Pull Request for Issue #24535 (part).

Replacement for #26469 (part). 

### Summary of Changes

This Pull Request (PR) fixes all datetime columns of database table `#__users` so there will not be any `Invalid value '0000-00-00 00:00:00' for datetime` error anymore on MySQL 5.7 or later when strict mode is enabled.

The `registerDate` is handled like columns `created` or `created_time` at other places with other recently merged PR's for datetime columns.

The `lastvisitDate` column was subject of a discussion in the first PR I made for com_users for that purpose, #26469 , and also oin Glip, and finally it turned out that we can remove the reference for the `lastvisitDate` column to `#__ucm_content`, what this PR here does, too, and so can make this column nullable, which is more logical than like it was in PR #26469 .

The `lastResetTime` (time of last password reset) column of the users table will also allow null values.

The user notes part which was part of PR #26469 has been separated into a separate PR, #26609 , which meanwhile has been merged, so user notes are not handled in this PR here.

### Testing Instructions

Testers please report back the database kind (MySQL or PostgreSQL) on which you have tested.

If you have both MySQL and PostgreSQL, please test on both if possible.

**Test 1: New installation**

1. Apply the patch on a clean 4.0-dev branch using git or merging manually, or apply it on an installation of current 4.0-dev using patchtester.
2. If used patchtester on an existing installation in step 1, delete file `configuration.php` and delete all Joomla database tables in PhpMyAdmin or PhpPgAdmin (depending on your database type).
3. Do a new installation, login to backend, confirm the statistics dialog, go to global config and set error reporting to maximum or development in server settings.
4. Play around with users, create some, modify some, use all possible options for activation, check the list display and its filters and sorting.
5. After any action in step 4, check in your database the relevant `datetime`/`timestamp without timezone` columns.

Result: See section "Expected result" below.

**Test 2: Update sql script**

1. Install a clean clean 4.0-dev, login to backend, confirm the statistics dialog, go to global config and set error reporting to maximum or development in server settings.
2. Apply the changes from this PR e.g. manually or with patch tester.
3. Open PhpMyAdmin or PhpPgAdmin (depending on your database type), select your database and then go to the SQL command input.
4. On MySQL copy the first line of file `installation/sql/mysql/joomla.sql` into the SQL command window but don't execute the commands yet:
```
SET SQL_MODE = "NO_AUTO_VALUE_ON_ZERO";
```
This switches off strict mode to the SQL will run on MySQL 5.7 or later.

5. Copy the content of file `administrator/components/com_admin/sql/updates/mysql/4.0.0-2019-10-17.sql` or `administrator/components/com_admin/sql/updates/postgresql/4.0.0-2019-10-17.sql` (depending on your database type) into the SQL command window, in case of MySQL below the previously pasted commands, but don't execute the commands yet.
7. Replace `#__` by your database prefix in the SQL statements pasted before in the SQL input window.
6. Put the cursor to the beginning of the 1st SQL statement in the SQL input window and now execute all SQL commands.
7. Play around with users, create some, modify some, use all possible options for activation, check the list display and its filters and sorting.
8. After any action in step 7, check in your database the relevant `datetime`/`timestamp without timezone` columns.

Result: See section "Expected result" below.

### Expected result

**New Installation**

Users work as well as without this PR. In a MySQL database there are no columns of data type `datetime` having value '0000-00-00 00:00:00' in table `#__users`, and there is no invalid default value anymore in MySQL >= 5.7 with strict mode on. On PostgreSQL there are no columns of data type `timestamp without time zone` having value '0000-00-00 00:00:00' in this table.

**Update sql script**

The statements are processed without error. The expected result is the same as for a new installation.

### Actual result

**New Installation**

On MySQL same as expected, but the default value '0000-00-00 00:00:00' of database columns of data type `datetime` is invalid in MySQL >= 5.7 with strict mode on, and there might be values '0000-00-00 00:00:00'. On postgreSQL same as expected, but there might be values '1970-01-01 00:00:00' in this table.

### Documentation Changes Required

Maybe core developer docs and extension developer docs should be updated to encourage them not to use '0000-00-00 00:00:00' on MySQL anymore but use real NULL and not abuse '1970-01-01 00:00:00' on PostgreSQL as a speudo null date anymore and use real NULL values also there.
